algorithms/unit_tests/TestSortByKey.hpp

```diff
@@ -0,0 +1,177 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_ALGORITHMS_UNITTESTS_TEST_SORT_BY_KEY_HPP
+#define KOKKOS_ALGORITHMS_UNITTESTS_TEST_SORT_BY_KEY_HPP
+
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+#include <Kokkos_Random.hpp>
+#include <Kokkos_Sort.hpp>
+
+namespace Test {
+namespace SortImpl {
+
+struct Less {
+  template <class ValueType>
+  KOKKOS_INLINE_FUNCTION bool operator()(const ValueType &lhs,
+                                         const ValueType &rhs) const {
+    return lhs < rhs;
+  }
+};
+
+struct Greater {
+  template <class ValueType>
+  KOKKOS_INLINE_FUNCTION bool operator()(const ValueType &lhs,
+                                         const ValueType &rhs) const {
+    return lhs > rhs;
+  }
+};
+
+template <class ExecutionSpace, class Keys, class Permute,
+          class Comparator = Less>
+struct is_sorted_by_key_struct {
+  Keys keys;
+  Keys keys_orig;
+  Permute permute;
+  Comparator comparator;
+
+  is_sorted_by_key_struct(Keys keys_, Keys keys_orig_, Permute permute_,
+                          Comparator comparator_ = Comparator{})
+      : keys(keys_),
+        keys_orig(keys_orig_),
+        permute(permute_),
+        comparator(comparator_) {}
+  KOKKOS_INLINE_FUNCTION
+  void operator()(int i, unsigned int &count) const {
+    if (i < keys.extent_int(0) - 1 && comparator(keys(i + 1), keys(i))) ++count;
+    if (keys(i) != keys_orig(permute(i))) ++count;
+  }
+};
+
+template <typename ExecutionSpace, typename ViewType>
+void iota(ExecutionSpace const &space, ViewType const &v,
```

Comment:
I think we have an Iota in algorithms. You should rename the kernel though if you implement it here :)

Reply:
I am not aware of `iota` in std algorithms. There is a broader discussion about implementing it in some form in #4498, but currently there is no resolution.
Given that this is in the test, I don't feel like it matters much what it's called.

Reply:
You're right, I could have sworn we discussed it a lot but I guess it never happened